File: spring-cloud-gateway/src/main/kotlin/info/novatec/gateway/springcloudgateway/AuthorizationFilterFactory.java
package info.novatec.gateway.springcloudgateway;

import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.factory.GatewayFilterFactory;
import org.springframework.http.HttpCookie;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.tuple.Tuple;

public class AuthorizationFilterFactory implements GatewayFilterFactory {

    @Override
    public GatewayFilter apply(Tuple args) {
        return (exchange, chain) -> {
            HttpCookie cookie = exchange.getRequest().getCookies().getFirst("customer-Id");
            if (cookie == null || cookie.getValue().isEmpty()) {
                ServerHttpResponse response = exchange.getResponse();
                response.setStatusCode(HttpStatus.BAD_REQUEST);
                response.setComplete();
                return chain.filter(exchange.mutate().response(response).build());
            }
            ServerHttpRequest request = exchange.getRequest().mutate()
                    .header(HttpHeaders.AUTHORIZATION, cookie.getValue())
                    .build();
            return chain.filter(exchange.mutate().request(request).build());
        };
    }
}

File: spring-cloud-gateway/src/main/kotlin/info/novatec/gateway/springcloudgateway/AuthorizationFilterFactory.kt
package info.novatec.gateway.springcloudgateway

import info.novatec.gateway.springcloudgateway.AuthorizationFilterFactory.Config
import org.springframework.cloud.gateway.filter.GatewayFilter
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component

/**
 * Filter that extracts the `customer-Id` cookie and adds it's value to the
 * request's Authorization header.
 *
 * Note: this filter is used for the `/images` route which is defined in
 * the application.yml.
 */
@Component
class AuthorizationFilterFactory : AbstractGatewayFilterFactory<Config>(Config::class.java) {

    override fun apply(config: Config): GatewayFilter {
        return GatewayFilter { exchange, chain ->
            val cookie = exchange.request.cookies.getFirst("customer-Id")
            if (cookie?.value.isNullOrEmpty()) {
                exchange.response.statusCode = HttpStatus.BAD_REQUEST
                exchange.response.setComplete()
            } else {
                val request = exchange.request
                        .mutate()
                        .header(HttpHeaders.AUTHORIZATION, cookie!!.value)
                        .build()
                chain.filter(exchange.mutate().request(request).build())
            }
        }
    }

    class Config
}
